lnd_interceptor.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/lnrpc/routerrpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+type LndHtlcInterceptor struct {
+	client        *LndClient
+	stopRequested bool
+}
+
+func NewLndHtlcInterceptor(client *LndClient) *LndHtlcInterceptor {
+	return &LndHtlcInterceptor{
+		client: client,
+	}
+}
+
+func (i *LndHtlcInterceptor) Start() error {
+	go forwardingHistorySynchronize(i.client)
+	go channelsSynchronize(i.client)
+	return i.intercept()
+}
+
+func (i *LndHtlcInterceptor) Stop() {
+	i.stopRequested = true
+}
+
+func (i *LndHtlcInterceptor) intercept() error {
+	for {
+		if i.stopRequested {
+			return nil
+		}
+
+		cancellableCtx, cancel := context.WithCancel(context.Background())
```

Comment:
defer cancel()

Reply:
Added cancel on all code paths. We're in a for loop, so now the context is cancelled on each iteration.